Geometry_Engine/Compute/OutlinesFromLines.cs

```diff
@@ -0,0 +1,427 @@
+            if (nodeValence == 0)
+                nodesByValence.Remove(point);
+            if (nodeValence == 1)
+            {
+                int index = lines.FindIndex(x => x.Start == point || x.End == point);
+                Line l = lines[index];
+                lines.RemoveAt(index);
+                nodesByValence.Remove(point);
+
+                if (l.Start == point)
+                {
+                    nodesByValence[l.End]--;
+                    l.End.CullValenceOne(nodesByValence, lines, distanceTolerance);
+                }
+                else
+                {
+                    nodesByValence[l.Start]--;
+                    l.Start.CullValenceOne(nodesByValence, lines, distanceTolerance);
+                }
+            }
+        }
+
+        /***************************************************/
+
+        private static TransformMatrix TransformToGlobalXY(this List<Line> lines)
+        {
+            Vector dir1 = lines[0].Direction();
+            Vector dir2 = lines.FirstOrDefault(x => x.Direction().IsParallel(dir1) == 0)?.Direction();
+            if (dir2 == null)
+                return null;
+
+            Vector normal = dir1.CrossProduct(dir2);
+            Cartesian local = Create.CartesianCoordinateSystem(lines[0].Start, dir1, dir1.CrossProduct(normal));
+            return local.OrientationMatrix(new Cartesian());
+        }
+
+        /***************************************************/
+
+        private static Output<Polyline, List<Line>> OuterAndInnerLines(this List<Line> lines, double distanceTolerance)
+        {
+            // Transform the lines to global XY plane
+            TransformMatrix toGlobal = lines.TransformToGlobalXY();
+            if (toGlobal == null)
+            {
+                BH.Engine.Base.Compute.RecordWarning("Some of the lines have been ignored in the process of creating outlines because they were all collinear.");
+                return null;
+            }
+
+            List<Line> transformed = lines.Select(x => x.Transform(toGlobal)).ToList();
+
+            // Turn the lines into a graph represented as a dictionary of nodes and adjoining edges
+            Dictionary<Point, List<Line>> graph = new Dictionary<Point, List<Line>>();
+            foreach (Line line in transformed)
+            {
+                if (!graph.ContainsKey(line.Start))
+                    graph.Add(line.Start, new List<Line>());
+
+                if (!graph.ContainsKey(line.End))
+                    graph.Add(line.End, new List<Line>());
+
+                graph[line.Start].Add(line);
+                graph[line.End].Add(line);
+            }
+
+            // Start at leftmost node
+            Point leftmost = graph.Keys.First();
+            foreach (Point node in graph.Keys)
+            {
+                if (node.X < leftmost.X)
+                    leftmost = node;
+            }
+
+            // Find edge starting at start node with leftmost dir
```

Comment:
There is no Y check in this for loop and there had never been 😉 We are looking for the leftmost only point, purely X coordinate based exercise.

Reply:
Line 223 says `currentDir.Y < dir.Y` - should that say `X` then? I know the above for loop on line 210 was handling `X` but this for loop looks like it's handling `Y`?

Reply:
Hmm, you may be right that I used a mental shortcut here while typing the code comment. Let's make sure we both are on the same page with the process itself:
1. We find leftmost node, i.e. a point lying at the left edge of the bounds
2. From that point, we find the direction looking leftmost in relation to the global Y - since all edges are on the right from the global Y (because we start from the leftmost node), the one with highest Y component will meet the description

So maybe `Find edge starting at leftmost start node with dir closest to global Y` would be more precise?

Reply:
I think that would make slightly more sense based on the code around the comment to me 👍